jenkins.sh

```diff
@@ -1,14 +1,57 @@
-#!/bin/bash -x
-
-# This removes rbenv shims from the PATH where there is no
-# .ruby-version file. This is because certain gems call their
-# respective tasks with ruby -S which causes the following error to
-# appear: ruby: no Ruby script found in input (LoadError).
-if [ ! -f .ruby-version ]; then
-  export PATH=$(printf $PATH | awk 'BEGIN { RS=":"; ORS=":" } !/rbenv/' | sed 's/:$//')
-fi
+#!/bin/bash
+
+GH_STATUS_REPO_NAME=${INITIATING_REPO_NAME:-"alphagov/imminence"}
+CONTEXT_MESSAGE=${CONTEXT_MESSAGE:-"default"}
+GH_STATUS_GIT_COMMIT=${INITIATING_GIT_COMMIT:-${GIT_COMMIT}}
+
+function github_status {
+  STATUS="$1"
+  MESSAGE="$2"
+  gh-status "$GH_STATUS_REPO_NAME" "$GH_STATUS_GIT_COMMIT" "$STATUS" -d "Build #${BUILD_NUMBER} ${MESSAGE}" -u "$BUILD_URL" -c "$CONTEXT_MESSAGE" >/dev/null
+}
+
+function error_handler {
+  trap - ERR # disable error trap to avoid recursion
+  local parent_lineno="$1"
+  local message="$2"
+  local code="${3:-1}"
+  if [[ -n "$message" ]] ; then
+    echo "Error on or near line ${parent_lineno}: ${message}; exiting with status ${code}"
+  else
+    echo "Error on or near line ${parent_lineno}; exiting with status ${code}"
+  fi
+  github_status error "errored on Jenkins"
+  exit "${code}"
+}
+
+trap 'error_handler ${LINENO}' ERR
+github_status pending "is running on Jenkins"
+
+# Cleanup anything left from previous test runs
+git clean -fdx
+
+# Try to merge master into the current branch, and abort if it doesn't exit
+# cleanly (ie there are conflicts). This will be a noop if the current branch
+# is master.
+git merge --no-commit origin/master || git merge --abort
```

Comment:
This is straight from the source, but I am a wee bit conflicted about it.  My understanding is that if you have a branch behind master, and it merges cleanly, you will be running the tests for your branch against something that's not actually the code in your branch.  I do think it's useful, but it seems like the kind of hidden option that will cause confusion.  (e.g. I run the specs on my machine, all ok, I push my branch, it runs on CI and breaks because I'm behind master, I'm all "but it worked on my machine. wait, that test isn't even in my branch, ...").

I do think it's useful to know if our branch merges well with master, and if it still passes test when it does so, but not at the expense of knowing if the tests I'm working on in my branch work correctly.  I wonder if it would it be preferable to parameterise this and have our branches build run the script twice: once to check the branch, then again to check the integration of the branch with master.

FWIW: I'm fairly sure this is what you get from Travis if you have both `build push` and `build pr` toggles set; one build for the branch (the push build) and one build for the merge (the pr build).